src/withData.tsx

```diff
@@ -1,19 +1,110 @@
 import React, { useMemo } from 'react'
 import Head from 'next/head'
+import { NextPageContext, NextPage } from 'next'
 import { ApolloProvider } from '@apollo/react-hooks'
-import { ApolloClient, InMemoryCache, HttpLink } from 'apollo-boost'
-import fetch from 'isomorphic-unfetch'
+import { ApolloClient, InMemoryCache, ApolloClientOptions } from 'apollo-boost'
+import { ApolloCache } from 'apollo-cache'
+import 'isomorphic-unfetch' // Use as a global polyfill, see https://www.npmjs.com/package/isomorphic-unfetch#usage
 import { isFunction } from 'lodash'
 
-let apolloClient = null
+type ApolloCacheShape = any
+type CreateCacheFunction = () => ApolloCache<ApolloCacheShape>
+type NextApolloClientOptions = ApolloClientOptions<ApolloCacheShape> & {
+  createCache?: CreateCacheFunction | null
+}
+type NextApolloConfigFunction = (
+  ctx?: NextPageContext
+) => NextApolloClientOptions
+type NextApolloConfig =
+  | NextApolloClientOptions
+  | ApolloClient<ApolloCache<ApolloCacheShape>>
+  | NextApolloConfigFunction
+
+type NextPageContextWithApollo = NextPageContext & {
+  apolloClient: ApolloClient<ApolloCacheShape>
+}
+
+let apolloClient: ApolloClient<ApolloCache<ApolloCacheShape>>
+
+const createDefaultCache: CreateCacheFunction = () => new InMemoryCache()
+
+/**
+ * Creates and configures the ApolloClient
+ * @param  {Object} [apolloConfig]
+ * @param  {Object} [initialState={}]
+ */
+function createApolloClient(
+  apolloConfig: NextApolloClientOptions,
+  initialState: ApolloCacheShape = {}
+): ApolloClient<ApolloCache<ApolloCacheShape>> {
+  const createCache = apolloConfig.createCache || createDefaultCache
+
+  const config = {
+    ssrMode: typeof window === 'undefined', // Disables forceFetch on the server (so queries are only run once)
+    cache: createCache().restore(initialState || {}),
+    ...apolloConfig
+  }
+
+  delete config.createCache
+
+  return new ApolloClient(config)
+}
+
+/**
+ * Always creates a new apollo client on the server
+ * Creates or reuses apollo client in the browser.
+ * @param  {Object} initialState
+ */
+function initApolloClient(
+  nextApolloConfig: NextApolloConfig,
```

Comment:
I made `NextApolloConfig` a distinct type so that we can accept an apollo-client config object (plus `createCache`), an `ApolloClient Instance`, or a function that accepts next's `ctx` and produces an apollo-client config object.